frontend/src/features/nextcloud_integration/components/FileSelector.tsx

```diff
@@ -0,0 +1,151 @@
+import { FileStat, WebDAVClient } from 'webdav';
+
+type FileSelectorProps = {
+  /** path to the directory starting from the users root */
+  path: string;
+  /** fires when an element is selected */
+  onSelect: (item: FileStat) => void;
+};
+
+const matchesSearchTerm = (searchTerm: string) => (file: FileStat) => {
+  const regex = new RegExp(searchTerm.toLowerCase());
+  return regex.test(file.basename.toLowerCase());
+};
+
+const sortByNameAsc = (a: FileStat, b: FileStat) => {
+  const nameA = a.basename.toLowerCase();
+  const nameB = b.basename.toLowerCase();
+  return nameA < nameB ? -1 : nameB < nameA ? 1 : 0;
+};
+
+const sortByNameDsc = (a: FileStat, b: FileStat) => {
+  const nameA = a.basename.toLowerCase();
+  const nameB = b.basename.toLowerCase();
+  return nameA > nameB ? -1 : nameB > nameA ? 1 : 0;
+};
+
+const sortByDateAsc = (a: FileStat, b: FileStat) => {
+  const dateA = new Date(a.lastmod);
+  const dateB = new Date(b.lastmod);
+  return dateA < dateB ? -1 : dateB < dateA ? 1 : 0;
+};
+
+const sortByDateDsc = (a: FileStat, b: FileStat) => {
+  const dateA = new Date(a.lastmod);
+  const dateB = new Date(b.lastmod);
+  return dateA > dateB ? -1 : dateB > dateA ? 1 : 0;
+};
+
+interface SortMethods {
+  [key: string]: (a: FileStat, b: FileStat) => number;
+}
+const sortMethods: SortMethods = {
+  name_asc: sortByNameAsc,
+  name_dsc: sortByNameDsc,
+  date_asc: sortByDateAsc,
+  date_dsc: sortByDateDsc,
+};
+
+const sortFiles = (attribute: 'name' | 'date', order: 'asc' | 'dsc') => {
+  const key: string = attribute + '_' + order;
+  return sortMethods[key];
+};
+
+/** component to select a file from the top level of a given Nextcloud directory */
+export const FileSelector = (props: FileSelectorProps) => {
+  const { path, onSelect } = props;
+  const webdav = useNextcloudWebDavClient();
+  const [searchTerm, setSearchTerm] = useState<string>('');
+  const [sortAttribute, setSortAttribute] = useState<'name' | 'date'>('name');
+  const [sortOrder, setSortOrder] = useState<'asc' | 'dsc'>('asc');
+
+  const { t } = useTranslation(['fileSelector']);
+
+  const { data, refetch, isLoading } = useQuery(['files', path], {
+    queryFn: () => (webdav as WebDAVClient).getDirectoryContents('/remote.php/webdav/' + path),
+    meta: {
+      errorMessage: t('fileSelector:error'),
+    },
+    staleTime: 10000,
+    refetchOnWindowFocus: false,
+    enabled: !!webdav && !!path,
+  });
+
+  if (isLoading) {
+    return <LoadingSpinner />;
+  }
+
+  return (
+    <div className="w-full max-w-2xl overflow-y-auto p-8">
+      <div className="flex items-end gap-4">
+        <div className="flex-1">
+          <SimpleFormInput
+            id="FileInput"
+            labelText={t('fileSelector:search')}
+            onChange={(e) => setSearchTerm(e.target.value)}
+          />
+        </div>
+        <div className="w-32">
+          <UploadFile
+            path={path}
+            onSuccess={() => {
+              refetch();
+            }}
```

Comment:
Rather than this, call `queryClient.invalidateQueries` in the onSuccess handler of the mutation.
```ts
onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [...] })
},
```